### PR TITLE
CHE-2463 Improve Panels behaviors in IDE

### PR DESCRIPTION
**Pull request by vitaliy-guliy**
Add the ability to maximize any panel.

https://github.com/eclipse/che/issues/2463

See video for details
https://youtu.be/xSp7lhgF4as

### CODE REVIEW

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/parts/PartStack.java [hunk 1]
```diff
@@ -20,10 +20,42 @@
...
+        /**
+         * The default state when the part stack is visible.
+         * Part stack can be minimized, maximized or collapsed.
+         */
+        NORMAL,
+
+        /**
+         * Part stack is minimized by minimize button.
+         * Having this state part stack can not be maximized or collapsed but
+         *  only can be restored by clicking the pab button.
```
Comment (ashumilova) [inline ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/parts/PartStack.java]: the pab button -> the tab button

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/parts/PartStack.java [hunk 2]
```diff
@@ -69,11 +101,34 @@
...
+    void collapse();
+
+    /**
+     * Minimizes / hides the part stack.
+     * The part state will not be retored when restoring the perspective state.
+     */
+    void minimize();
+
+    /**
+     * Restores the part stack ant the perspective to the default state.
```
Comment (ashumilova) [inline ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/parts/PartStack.java]: ant -> and

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/parts/base/BasePresenter.java [hunk 3]
```diff
@@ -24,20 +24,28 @@
 
     protected PartStack partStack;
 
-    protected BasePresenter() {
+    @Override
+    public void onMaximize() {
```
Comment (ashumilova) [inline ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/parts/base/BasePresenter.java]: onMaximize -> maybe smth like onToogleMaximize() ? cause will do restore or maximize depending on current state

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/parts/PartStack.java (hunk 1 above)
Comment (vitaliy-guliy) [reply ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/parts/PartStack.java]: Thanks

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/parts/PartStack.java (hunk 2 above)
Comment (vitaliy-guliy) [reply ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/parts/PartStack.java]: Thanks

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/parts/base/BasePresenter.java (hunk 3 above)
Comment (vitaliy-guliy) [reply ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/parts/base/BasePresenter.java]: The method is renamed. Thanks

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/FontAwesome.java [hunk 4]
```diff
@@ -8,7 +8,7 @@
  * Contributors:
  *   Codenvy, S.A. - initial API and implementation
  *******************************************************************************/
-package org.eclipse.che.ide.ui;
+package org.eclipse.che.ide.api;
```
Comment (azatsarynnyy) [inline ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/FontAwesome.java]: Why have you moved _FontAwesome_ class from the **che-core-ide-ui** to the **che-core-ide-api** module?

Comment (azatsarynnyy) [reply ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/FontAwesome.java]: I think **che-core-ide-ui** module is the right place for that. And this class isn't a part of IDE API.

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/parts/PartStack.java [hunk 5]
```diff
@@ -95,4 +150,27 @@
...
+        /**
+         * Requests the delegate to maximize the part stack.
+         *
+         * @param partStack
+         *          part stack
+         */
+        void onMaximize(PartStack partStack);
+
+        /**
+         * Requests the delegate to  restore part stack state,
```
Comment (azatsarynnyy) [inline ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/parts/PartStack.java]: please replace comma with dot at the end

**Comment by azatsarynnyy** [comment]
A lot of tests have removed.
But no new tests for the new functionality. :scream_cat:

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/FontAwesome.java (hunk 4 above)
Comment (vitaliy-guliy) [reply ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/FontAwesome.java]: Yes, sure. But it's used in 
org.eclipse.che.ide.api.parts.base.BaseView

**Comment by vitaliy-guliy** [comment]
Three tests have been removed as functionality they are testing already is not exist.
Ads added few tests for the new functionality.

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/FontAwesome.java (hunk 4 above)
Comment (azatsarynnyy) [reply ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/FontAwesome.java]: ok. I see. Thanks

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/part/PartStackPresenter.java [hunk 6]
```diff
@@ -273,19 +271,123 @@ public void setFocus(boolean focused) {
...
+        if (delegate != null) {
+            delegate.onMaximize(this);
+        }
+    }
+
+    @Override
+    public void collapse() {
+        view.setMaximized(false);
+
+        if (state != State.NORMAL) {
```
Comment (vparfonov) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/part/PartStackPresenter.java:307]: Can you explain behavior here?

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/part/PartStackPresenter.java [hunk 7]
```diff
@@ -273,19 +271,123 @@ public void setFocus(boolean focused) {
...
+    @Override
+    public State getPartStackState() {
+        return state;
+    }
+
+    @Override
+    public void minimize() {
+        view.setMaximized(false);
+
+        if (state == State.MAXIMIZED) {
```
Comment (vparfonov) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/part/PartStackPresenter.java:333]: Can you explain behavior here?

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/part/editor/EditorPartStackView.java [hunk 8]
```diff
@@ -248,6 +250,11 @@ public void setFocus(boolean focused) {
         }
     }
 
+    @Override
+    public void setMaximized(boolean maximized) {
+        getElement().setAttribute("maximized", "" + maximized);
```
Comment (vparfonov) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/part/editor/EditorPartStackView.java]: Boolean.toString() ?

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/part/editor/multipart/EditorMultiPartStackPresenter.java [hunk 9]
```diff
@@ -138,11 +143,28 @@ public void setActivePart(@NotNull PartPresenter part) {
     }
 
     @Override
-    public void hidePart(PartPresenter part) {
-        EditorPartStack editorPartStack = getPartStackByPart(part);
-        if (editorPartStack != null) {
-            editorPartStack.hidePart(part);
-        }
+    public void maximize() {
```
Comment (vparfonov) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/part/editor/multipart/EditorMultiPartStackPresenter.java:146]: What the reason change state but not change view?

Comment (vitaliy-guliy) [reply ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/part/editor/multipart/EditorMultiPartStackPresenter.java:146]: Actually the central part stack become maximized when right, left and bottom part stacks are collapsed. So, it's just to store the part stack state.

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/part/PartStackPresenter.java (hunk 7 above)
Comment (vitaliy-guliy) [reply ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/part/PartStackPresenter.java:333]: Please see the following commit.

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/part/editor/EditorPartStackView.java (hunk 8 above)
Comment (vitaliy-guliy) [reply ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/part/editor/EditorPartStackView.java]: Done. Thanks

**Review by vparfonov — APPROVED** [review]

**Comment by vitaliy-guliy** [comment]
FontAwesome has been moved to ide/commons-gwt ( che-core-commons-gwt )

**Review by vparfonov — APPROVED** [review]

**Review by ashumilova — APPROVED** [review]

**Pull request merged**